File: app/src/main/java/com/kieran/winnipegbusbackend/FavouriteStop.java
package com.kieran.winnipegbusbackend;

import com.google.android.gms.maps.model.LatLng;

import java.io.Serializable;

public class FavouriteStop extends Stop implements Serializable{
    private int timesUsed;
    private transient LatLng latLng; //TODO seperate this somehow?
    private String alias;

    public FavouriteStop(String stopName, int stopNumber, int timesUsed) {
        super(stopName, stopNumber);
        this.timesUsed = timesUsed;
    }

    public FavouriteStop(String stopName, int stopNumber) {
        super(stopName, stopNumber);
        this.timesUsed = 0;
    }

    public int getTimesUsed() {
        return timesUsed;
    }

    public void use() {
        timesUsed++;
    }

    public LatLng getLatLng() {
        return latLng;
    }

    public void setLatLng(LatLng latLng) {
        this.latLng = latLng;
    }

    public String getAlias() {
        return alias;
    }

    public void setAlias(String alias) {
        if(alias.equals(stopName))
            this.alias = null;
        else
            this.alias = alias;
    }

    public String getDisplayName() {
        return alias == null ? stopName : alias;
    }
}

File: app/src/main/java/com/kieran/winnipegbusbackend/common/FavouriteStop.kt
package com.kieran.winnipegbusbackend.common

import com.kieran.winnipegbusbackend.interfaces.RouteIdentifier
import com.kieran.winnipegbusbackend.interfaces.StopIdentifier

import java.io.Serializable

class FavouriteStop : Stop, Serializable {
    var timesUsed: Int = 0
        private set
    var alias: String? = null
        set(alias) = if (alias == name)
            field = null
        else
            field = alias

    override val displayName: String
        get() = if (this.alias == null) name else this.alias!!

    constructor(stopName: String, stopIdentifier: StopIdentifier, timesUsed: Int) : super(stopName, stopIdentifier) {
        this.timesUsed = timesUsed
    }

    constructor(stopName: String, stopIdentifier: StopIdentifier) : super(stopName, stopIdentifier) {
        this.timesUsed = 0
    }

    constructor(name: String, identifier: StopIdentifier, timesUsed: Int = 0, latlng: GeoLocation? = null, id: Long = -1, alias: String? = null, routes: Collection<RouteIdentifier>? = null) : super(name, identifier, latlng) {
        this.timesUsed = timesUsed
        this.id = id
        this.alias = alias
        this.routes = routes
    }

    fun use() {
        timesUsed++
    }
}
